spikeinterface/core/basesorting.py

```diff
@@ -76,10 +121,35 @@ def get_unit_spike_train(
         else:
             return spike_frames
 
-    def register_recording(self, recording):
+    def register_recording(self, recording, check_spike_frames=True):
+        """Register a recording to the sorting.
+
+        Parameters
+        ----------
+        recording : BaseRecording
+            Recording with the same number of segments as current sorting.
+            Assigned to self._recording.
+        check_spike_frames : bool, optional
+            If True, assert for each segment that all spikes are within the recording's range.
+            By default True.
+        """
         assert np.isclose(self.get_sampling_frequency(),
                           recording.get_sampling_frequency(),
                           atol=0.1), "The recording has a different sampling frequency than the sorting!"
+        assert self.get_num_segments() == recording.get_num_segments(), (
+            "The recording has a different number of segments than the sorting!"
+        )
+        if check_spike_frames:
+            for sorting_segment, rec_segment in zip(self._sorting_segments, recording._recording_segments):
+                for unit_id in self.unit_ids:
+                    unit_segment_spikes = sorting_segment.get_unit_spike_train(
+                        unit_id=unit_id, 
+                        start_frame=None,
+                        end_frame=None,
+                    )
+                    segment_n_samples = rec_segment.get_num_samples()
+                    if any([spike_frame >= segment_n_samples for spike_frame in unit_segment_spikes]):
+                        raise ValueError("Sorting's spike trains are inconsistent with the recording's duration.")
```

Comment:
@TomBugnon we found that sometimes KS returns spikes times beyond the duration of the recording. This is clearly wrong, but we implemented a curation function to [`remove_excess_spikes`](https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/curation/remove_excess_spikes.py#L60).

When extracting waveforms, we [check](https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/waveform_extractor.py#L248-L252) that the sorting is consistent with the recording in this sense.

In my opinion, we should raise a warning here, so that the user can still register the recording and then remove the excess spikes. What do you think?

Reply:
I see. If you think it should be possible to register a shorter recording then we probably want to call has_excess_spikes and raise an exception both in concatenate_sortings and frameslicesorting, which may rely on the registered recording (instead of only once here)